### PR TITLE
Chain-in-chord that fails before the last task in the chain now triggers chord error handling

### LINKED ISSUE #6220: Chain-in-chord that fails before the last task in the chain doesn't trigger chord error handling
# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [ ] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [x] I have tried reproducing the issue on more than one workers pool.
- [x] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [x] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.
Confirmed to exist on at least celery 4.3.0 as well.

#### Related Issues

- #3709  is related to error handling for chords, though this is more of an edge case in which the behavior is problematic, rather than just misalignment of docs and actual behavior.

#### Possible Duplicates

- There are several other issues that deal with chains-in-chords, e.g. #5229 seems similar, but nothing that addresses this exact problem. I don't think this duplicates them, but I haven't dug into the celery code enough to be sure that they aren't caused the the same underlying misbehavior in the redis backend.

## Environment & Settings

**Celery version**: 4.4.6 (cliffs)

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:4.4.6 (cliffs) kombu:4.6.11 py:3.6.9
            billiard:3.6.3.0 redis:3.5.3
platform -> system:Linux arch:64bit, ELF
            kernel version:4.4.0-17763-Microsoft imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis://localhost:6379/

broker_url: 'redis://localhost:6379//'
result_backend: 'redis://localhost:6379/'
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

Not sure that any of these are minimal, but I'll include what I can test locally:
* **Minimal Python Version**: 3.6
* **Minimal Celery Version**: 4.3
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: 6.0.4
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==2.6.0
billiard==3.6.3.0
celery==4.4.6
future==0.18.2
importlib-metadata==1.7.0
kombu==4.6.11
pkg-resources==0.0.0
pytz==2020.1
redis==3.5.3
vine==1.3.0
zipp==3.1.0
```

</p>
</details>

### Other Dependencies

<details>
<p>

Output of `redis-server --version`, in case it's useful:
```
Redis server v=6.0.4 sha=00000000:0 malloc=jemalloc-5.1.0 bits=64 build=d5d470262d47f1a0
```

</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>
Basic test case below. I'm still working on converting this to an integration test for your test suite.

```python
from celery import Celery

app = Celery('task_fail', backend='redis://localhost:6379', broker='redis://localhost:6379')

@app.task
def simple_pass():
    print("Passing")
    pass

@app.task
def simple_fail():
    print("Failing!")
    raise Exception('Fail!')

@app.task
def record_failure(*args, **kwargs):
    print(f'Fail with args={args}, kwargs={kwargs}')

if __name__ == "__main__":
    from celery import chord

    # Test 1: Passes - When the chain fails at the end, the failure logic on
    # the chord is executed.
    try:
        chain = simple_pass.si() | simple_fail.si()
        ~chord([simple_pass.si(), chain],
                simple_pass.si().on_error(record_failure.s(test="test_1")))
    except Exception:
        print("Test 1 worked")

    # Test 2: Fails - When the chain has an intermediate exception, the failure
    # logic on the chord is not executed & the client simply hangs.
    try:
        chain = simple_pass.si() | simple_fail.si() | simple_pass.si()
        ~chord([simple_pass.si(), chain],
                simple_pass.si().on_error(record_failure.s(test="test_2")))
    except Exception:
        print("Test 2 worked")
```
Note that the only difference between test 1 and test 2 is that test 2 add the `simple_pass.si()` call to the end of the chain. That seems to be enough to trigger this error.
</p>
</details>

# Expected Behavior
Test 2 should behave the same way that Test 1 does. Output of Test 1, with some simplification:
```
[2020-07-10 15:28:52,940: INFO/MainProcess] Received task: task_fail.simple_pass[65241973-d598-4bbe-a0c8-1f86694c9f33]
[2020-07-10 15:28:52,945: WARNING/ForkPoolWorker-8] Passing
[2020-07-10 15:28:52,954: INFO/ForkPoolWorker-8] Task task_fail.simple_pass[65241973-d598-4bbe-a0c8-1f86694c9f33] succeeded in 0.008173699956387281s: None
[2020-07-10 15:28:52,967: INFO/MainProcess] Received task: task_fail.simple_pass[fc15c163-0616-4b87-94c3-2c1b225f2e0e]
[2020-07-10 15:28:52,972: WARNING/ForkPoolWorker-8] Passing
[2020-07-10 15:28:52,985: INFO/ForkPoolWorker-8] Task task_fail.simple_pass[fc15c163-0616-4b87-94c3-2c1b225f2e0e] succeeded in 0.013192899990826845s: None
[2020-07-10 15:28:52,985: INFO/MainProcess] Received task: task_fail.simple_fail[0b0a3925-7ea8-4423-8e87-3a9f00d5c967]
[2020-07-10 15:28:52,992: WARNING/ForkPoolWorker-8] Failing!
[2020-07-10 15:28:53,010: ERROR/ForkPoolWorker-8] Chord callback for '76840678-80a3-4915-96b8-04c135ecb237' raised: ChordError("Dependency 0b0a3925-7ea8-4423-8e87-3a9f00d5c967 raised Exception('Fail!',)",)
Traceback (most recent call last):
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "~/celery/task_fail.py", line 13, in simple_fail
    raise Exception('Fail!')
Exception: Fail!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/redis.py", line 449, in on_chord_part_return
    callback.delay([unpack(tup, decode) for tup in resl])
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/redis.py", line 449, in <listcomp>
    callback.delay([unpack(tup, decode) for tup in resl])
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/redis.py", line 402, in _unpack_chord_result
    raise ChordError('Dependency {0} raised {1!r}'.format(tid, retval))
celery.exceptions.ChordError: Dependency 0b0a3925-7ea8-4423-8e87-3a9f00d5c967 raised Exception('Fail!',)
[2020-07-10 15:28:53,022: INFO/MainProcess] Received task: task_fail.record_failure[5f264f5d-a330-4f01-9274-662474d31a8f]
[2020-07-10 15:28:53,026: ERROR/ForkPoolWorker-8] Task task_fail.simple_fail[0b0a3925-7ea8-4423-8e87-3a9f00d5c967] raised unexpected: Exception('Fail!',)
Traceback (most recent call last):
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "~/celery/task_fail.py", line 13, in simple_fail
    raise Exception('Fail!')
Exception: Fail![2020-07-10 15:28:53,028: WARNING/ForkPoolWorker-1] Fail with args=('bc3aeb74-68c1-40e9-9c55-be9fb7860786',), kwargs={'test': 'test_1'}
[2020-07-10 15:28:53,036: INFO/ForkPoolWorker-1] Task task_fail.record_failure[5f264f5d-a330-4f01-9274-662474d31a8f] succeeded in 0.007674199994653463s: None
```
The most important thing to notice here is that the `record_failure` task is called & logs the failure, including the ID of the failing task. I expect this behavior to occur in both cases.

# Actual Behavior

The output of the worker on Test 2 is:
```
[2020-07-10 15:28:55,039: INFO/MainProcess] Received task: task_fail.simple_pass[5779ce6a-64af-40ea-9b26-8ff3bdad348d]
[2020-07-10 15:28:55,050: INFO/MainProcess] Received task: task_fail.simple_pass[39469dfa-401c-45f0-9314-5849f6e7ca5f]
[2020-07-10 15:28:55,051: WARNING/ForkPoolWorker-8] Passing
[2020-07-10 15:28:55,051: WARNING/ForkPoolWorker-1] Passing
[2020-07-10 15:28:55,058: INFO/ForkPoolWorker-8] Task task_fail.simple_pass[5779ce6a-64af-40ea-9b26-8ff3bdad348d] succeeded in 0.007474799989722669s: None
[2020-07-10 15:28:55,062: INFO/ForkPoolWorker-1] Task task_fail.simple_pass[39469dfa-401c-45f0-9314-5849f6e7ca5f] succeeded in 0.010612400015816092s: None
[2020-07-10 15:28:55,069: INFO/MainProcess] Received task: task_fail.simple_fail[2fead196-cfe2-4ce5-896f-303c2f9970e2]
[2020-07-10 15:28:55,073: WARNING/ForkPoolWorker-8] Failing!
[2020-07-10 15:28:55,081: ERROR/ForkPoolWorker-8] Task task_fail.simple_fail[2fead196-cfe2-4ce5-896f-303c2f9970e2] raised unexpected: Exception('Fail!',)
Traceback (most recent call last):
  File "/home/crawjor/test/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 412, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/crawjor/test/celery/celery.venv/lib/python3.6/site-packages/celery/app/trace.py", line 704, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/crawjor/test/celery/task_fail.py", line 13, in simple_fail
    raise Exception('Fail!')
Exception: Fail!
```
Then no more progress is made. Note that neither the error handler nor the callback is ever invoked, despite the chain failing (which should trigger failure handlers in the chord).

The client side also hangs waiting on redis:
```
File "task_fail.py", line 40, in <module>
    simple_pass.si().on_error(record_failure.s(test="test_2")))
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/canvas.py", line 475, in __invert__
    return self.apply_async().get()
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/result.py", line 237, in get
    on_message=on_message,
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/asynchronous.py", line 200, in wait_for_pending
    for _ in self._wait_for_pending(result, **kwargs):
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/asynchronous.py", line 268, in _wait_for_pending
    on_interval=on_interval):
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/asynchronous.py", line 55, in drain_events_until
    yield self.wait_for(p, wait, timeout=interval)
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/asynchronous.py", line 64, in wait_for
    wait(timeout=timeout)
  File "~/celery/celery.venv/lib/python3.6/site-packages/celery/backends/redis.py", line 161, in drain_events
    message = self._pubsub.get_message(timeout=timeout)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/client.py", line 3617, in get_message
    response = self.parse_response(block=False, timeout=timeout)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/client.py", line 3503, in parse_response
    if not block and not conn.can_read(timeout=timeout):
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/connection.py", line 734, in can_read
    return self._parser.can_read(timeout)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/connection.py", line 321, in can_read
    return self._buffer and self._buffer.can_read(timeout)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/connection.py", line 231, in can_read
    raise_on_timeout=False)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/connection.py", line 198, in _read_from_socket
    data = recv(self._sock, socket_read_size)
  File "~/celery/celery.venv/lib/python3.6/site-packages/redis/_compat.py", line 72, in recv
```

There are no active, revoked, scheduled, or reserved tasks, and the queue is empty:
```
~/celery$ celery -A task_fail inspect active
-> celery: OK
    - empty -
~/celery$ celery -A task_fail inspect scheduled
-> celery: OK
    - empty -
~/celery$ celery -A task_fail inspect revoked
-> celery: OK
    - empty -
~/celery$ celery -A task_fail inspect reserved
-> celery: OK
    - empty -
```

### PR DESCRIPTION
**Pull request by msci-jordancrawford**
Marked with xfail for now.

*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://docs.celeryproject.org/en/master/contributing.html).

## Description

Fixes #6220.

### CODE REVIEW

**Comment by msci-jordancrawford** [comment]
I went ahead and added a fix as well, which should at least cause the tests to pass. That said, it really doesn't seem like the correct fix - if any of the maintainers have a suggestion for a more elegant way to fix this, it would be much appreciated. Thanks ahead of time! 

I'm also not sure what the 'correct' behavior here is - should the client be expecting a `ChordError` or an `ExpectedException`?

File: celery/backends/base.py [hunk 1]
```diff
@@ -170,6 +170,13 @@ def mark_as_failure(self, task_id, exc,
         if request:
             if request.chord:
                 self.on_chord_part_return(request, state, exc)
+            elif request.chain and 'chord' in request.chain[-1]['options']:
```
Comment (auvipy) [inline celery/backends/base.py:173]: does this solve the reported issue?

Comment (msci-jordancrawford) [reply celery/backends/base.py:173]: Yes. The problem is that this snippet of code is executed for the task in the chain that fails, but only the final task in the chain passes the `if request.chord` check, so the chord only gets notified if it's the final task in the chain that fails. Adding this check solves the problem as bluntly as possible: by simply checking for a chord associated with the last element in the chain and updating it if necessary. Very much unsure if that is the right way forward. 

It looks like it also breaks a couple of other tests. I think fixing those tests would just be a matter of changing this line to:
```python
elif getattr(request, 'chain', None) and 'chord' in request.chain[-1]['options']:
```
but that just further reinforces my idea that this is maybe not the right way/place to be fixing this problem.

Comment (auvipy) [reply celery/backends/base.py:173]: keep digging with it please

File: t/integration/test_canvas.py [hunk 2]
```diff
@@ -862,6 +862,24 @@ def test_chord_on_error(self, manager):
         assert len([cr for cr in chord_results if cr[2] != states.SUCCESS]
                    ) == 1
 
+    def test_chain_in_chord_on_error(self, manager):
```
Comment (auvipy) [inline t/integration/test_canvas.py:865]: the integration test has changed upstream, can you rebase, please?

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by thedrow** [comment]
@matusvalo Please rebase this. If you have the time, please debug this further.

**Comment by maxfenv** [comment]
I don't know whether this is best reported here or on the issue, but this seems to be where the bulk of the discussion is.  I hope the following is useful:

I've rebased this PR on top of the v4.4.7 tagged commit here: https://github.com/celery/celery/compare/v4.4.7...maxfenv:chord_in_chain_error_propogation

In the context of one of the projects we maintain that uses celery, I wrote some unit tests that highlight the issue in that context and that fail with celery 4.4.7 but pass with this pr applied. I also applied the change suggested by @msci-jordancrawford (`elif getattr(request, 'chain', None) and 'chord' in request.chain[-1]['options']:`). With that applied I do in fact get no test failures on this branch: https://github.com/maxfenv/celery/tree/chord_in_chain_error_propogation.  For avoidance of doubt, I'm running celery unit tests using docker like so:

```
cd docker
sudo docker-compose build celery
sudo docker-compose run --rm celery make test
```

For anyone wanting to replicate that exactly, it is also necessary to pin the dependencies in `requirements/dev.txt` to kombu<5 and amqp<5 for the docker image to build properly.

I hope to do a large scale integration test of this fix tomorrow and should be able to report on it's outcome by the day after that.

Thank you very much @msci-jordancrawford  for your work in hunting down, documenting and fixing this issue.  This has been causing us grief for some time!

**Comment by thedrow** [comment]
Could you please contribute the unit tests?

**Comment by maxfenv** [comment]
Yes I can do, though they are fundamentally not different to the test already included here.  I could commit them to my fork or include them as a snipped here, I of course can't push directly to the source branch of this PR.

**Comment by thedrow** [comment]
If they are not different, then why did you write them?

**Comment by maxfenv** [comment]
I wrote them because the barrier to entry was lower than figuring out how to run the celery project's unit tests from cold.  I've since done that anyway, but it seemed like the thing to do at the time.  Again, I wrote those tests in another project.

In any case, I've added a those hopefully non controversial test cases to `t/integration/test_canvas.py` alongside the test written by @msci-jordancrawford.  You may find them here:  https://github.com/maxfenv/celery/tree/chord_in_chain_error_propogation (exact commit: https://github.com/celery/celery/commit/1d73847b8b01f5ee44bcb1916ecbda407cbb073d).

I'm unsure about decorating them with `@flakey`, I suppose they should be, since they rely on the result returning before the timeout.

**Comment by maxfenv** [comment]
Unfortunately error handling appears to not work if the failing task is not in the last two tasks of a chain.  That is to say: The chord error fires correctly if the failing task in the chain is either the last or second to last, but not if it is (say) the 1st or 2nd out of 4.

Here are some more tests to highlight that issue: https://github.com/celery/celery/commit/1a3b36bf0b6bdfd740dfc88f21c1d7b1dd9964d7

We shall have a think about what the fix might look like and report back.  Since this is starting to diverge substantially from the commits in this PR at this point, would it be acceptable to open a new PR with the fix (if one is forthcoming).  I will still base the branch off of this one, so @msci-jordancrawford's commits remain in the history.  I don't want to claim any undue credit here.

**Comment by maxfenv** [comment]
The issue appears to be that the context with the 'chord' key in the 'options' is always the first element of the `request.chains` list.  As such, it should be sufficient to index like so: `request.chain[0]['options']`, instead of with the -1.  

However, I'm not certain on that point, so instead I opted to check for all contexts with a 'chord' and to pick the first among those, if any.  I've commited those changes to my version of this branch: https://github.com/maxfenv/celery/tree/chord_in_chain_error_propogation.  The code there passes all tests, apart from `t/unit/backends/test_database.py:209 test_DatabaseBackend.test_cleanup` for reasons that still escape me.

**Comment by maxfenv** [comment]
The test failure mention above only occurs on python < 3.8 and is, I believe, related to mismatching versions of pickle.  Something like this: https://stackoverflow.com/questions/63329657/python-3-7-error-unsupported-pickle-protocol-5

**Comment by maybe-sybr** [comment]
Oh my gosh, yes! Thank you for the changeset @msci-jordancrawford and thanks for bumping this @maxfenv . I was smashing my face against this all of yesterday without having seen this.

My gut feel is that this approach will work but I was a bit concerned about reaching deep into the `options` to find the appropriate group_id to fire a chord_part_complete for. The test cases will help and I'll have a solid half day to dig into this tomorrow :)

**Comment by maybe-sybr** [comment]
There is a PR up (#6746) which should fix this issue

**Pull request closed without merging**